sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/OfferEvaluator.java

```diff
@@ -0,0 +1,162 @@
+package com.mesosphere.sdk.offer.evaluate;
+
+import com.google.inject.Inject;
+import com.google.protobuf.TextFormat;
+import com.mesosphere.sdk.offer.*;
+import com.mesosphere.sdk.specification.PodInstance;
+import com.mesosphere.sdk.specification.ResourceSpecification;
+import com.mesosphere.sdk.specification.TaskSpec;
+import com.mesosphere.sdk.specification.VolumeSpecification;
+import com.mesosphere.sdk.state.StateStore;
+import com.mesosphere.sdk.state.StateStoreException;
+import org.apache.mesos.Protos.*;
+import com.mesosphere.sdk.scheduler.plan.PodInstanceRequirement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * The OfferEvaluator processes {@link Offer}s and produces {@link OfferRecommendation}s.
+ * The determination of what {@link OfferRecommendation}s, if any should be made are made
+ * in reference to the {@link OfferRequirement with which it was constructed.  In the
+ * case where an OfferRequirement has not been provided no {@link OfferRecommendation}s
+ * are ever returned.
+ */
+public class OfferEvaluator {
+    private static final Logger logger = LoggerFactory.getLogger(OfferEvaluator.class);
+
+    private final StateStore stateStore;
+    private final OfferRequirementProvider offerRequirementProvider;
+
+    @Inject
+    public OfferEvaluator(StateStore stateStore, OfferRequirementProvider offerRequirementProvider) {
+        this.stateStore = stateStore;
+        this.offerRequirementProvider = offerRequirementProvider;
+    }
+
+    public List<OfferRecommendation> evaluate(PodInstanceRequirement podInstanceRequirement, List<Offer> offers)
+            throws StateStoreException, InvalidRequirementException {
+        OfferRequirement offerRequirement = getOfferRequirement(podInstanceRequirement);
+        List<OfferEvaluationStage> evaluationStages = getEvaluationPipeline(podInstanceRequirement, offerRequirement);
+        List<OfferRecommendation> recommendations = Collections.emptyList();
+
+        for (int i = 0; i < offers.size(); ++i) {
+            if (!recommendations.isEmpty()) {
+                break;
+            }
+
+            Offer offer = offers.get(i);
+            MesosResourcePool resourcePool = new MesosResourcePool(offer);
+            OfferRecommendationSlate recommendationSlate = new OfferRecommendationSlate();
+            List<String> failureNotifications = new ArrayList<>();
+            try {
+                for (OfferEvaluationStage evaluationStage : evaluationStages) {
+                    evaluationStage.evaluate(
```

Comment:
I think I'll omit any explicit handling of the passing case, since each stage logs what passes itself, and it would complicate things unnecessarily (you'd have to pass the success notification string out in addition to whatever else, I don't think we should reify logging concerns to the point that they become part of the actual interface), but will add the names and fix the short-circuiting.